### PR TITLE
Add ESLint to Project

### PR DESCRIPTION
**Pull request by nicholaschuayunzhi**
Adds ESLint and its lodash plugin as a dev dependency. --fix command not executed yet. The eslintrc file was taken from https://github.com/MarkBind/markbind/blob/master/.eslintrc.js

### CODE REVIEW

**Comment by Gisonrg** [comment]
Nice one! Can't wait to fix all the style issues in the project :D

Before I merge this, do you think we should use an existing style standard for our eslint?
I am thinking about using either https://github.com/airbnb/javascript or https://github.com/google/eslint-config-google

Feel free to discuss with the team to choose a standard you all agreed on.

**Comment by nicholaschuayunzhi** [comment]
I've used airbnb before. I would prefer airbnb as their docs have the eslint rules which we can ctrl f easily to see an example of.

**Comment by damithc** [comment]
The common standards for our projects (open for updates) https://github.com/oss-generic/process/blob/master/docs/CodingStandards.adoc

**Comment by Gisonrg** [comment]
Let's use the airbnb preset and modified based on our standard then :P

 @nicholaschuayunzhi Can you help us set up the eslint as discussed?

**Comment by acjh** [comment]
> The common standards for our projects (open for updates) https:\//github.com/oss-generic/process/blob/master/docs/CodingStandards.adoc
>> If a topic is not covered in this document, refer to airbnb JavaScript Style Guide

airbnb it is! Also, [teammates-eslint.yml](https://github.com/TEAMMATES/teammates/blob/60a9efd3b5f1c7e1e0d035160bf9ed1f53e5d800/static-analysis/teammates-eslint.yml).

**Comment by nicholaschuayunzhi** [comment]
Okay I'll work on it. Need to do some research first.

Also do you want it to be standardized across all the modules?

**Comment by Gisonrg** [comment]
Nice, take you time. Let's start from `markbind-cli` first, then extend it to `markbind` (vue modules has different style standard).

**Comment by nicholaschuayunzhi** [comment]
Changed rules wrt to [oss-generic coding standards](https://docs.google.com/document/d/1gZ6WG6HBTJYHAtVkz9kzi_SUuzfXqzO-SvFnLuag2xM/pub?embedded=true)

Rules that can't be modified with existing eslint rules:
- Constants named in screaming snake case
- Prefer higher-level breaks to lower-level breaks. In the example below, the first is preferred, since the break occurs outside the parenthesized expression, which is at a higher level.
- suggestion for ternary expression
image

Read up on writing custom rules. Is it worth the time? Anyone with experience? Also will add to dev guide on how to set up eslint.

**Comment by nicholaschuayunzhi** [comment]
I've updated the eslint config as discussed. Disabled func-names rule and added line break style.

File: .eslintrc.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+module.exports = {
+  "env": {
+    "node": true,
+    "es6": true
+  },
+  "extends": "airbnb-base",
+  "rules": {
+    "function-paren-newline": "off",
+    "func-names": "off",
```
Comment (acjh) [inline .eslintrc.js]: Alphabetical (ascii) order:
- `func-names` before `function-paren-newline`
- `FunctionDeclaration` before `FunctionExpression`

File: .eslintrc.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    "indent": [
+      "error",
+      2,
+      {
+        "ArrayExpression": "first",
+        "CallExpression": { "arguments": "first" } ,
+        "FunctionExpression": { "parameters": "first" },
+        "FunctionDeclaration": { "parameters": "first" },
+        "MemberExpression": 1,
```
Comment (acjh) [inline .eslintrc.js]: `1` seems to be the default in [indent](https://eslint.org/docs/rules/indent.html) and is not overridden by [airbnb](https://github.com/airbnb/javascript).

**Review by acjh — CHANGES_REQUESTED** [review]

**Comment by acjh** [comment]
I took a while to understand why some choices were made.

Perhaps we should document this in the Wiki, together with the setup instructions.

File: .eslintrc.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    "func-names": "off",
+    "function-paren-newline": "off",
+    "indent": [
+      "error",
+      2,
+      {
+        "ArrayExpression": "first",
+        "CallExpression": { "arguments": "first" } ,
+        "FunctionExpression": { "parameters": "first" },
+        "FunctionDeclaration": { "parameters": "first" },
```
Comment (acjh) [inline .eslintrc.js]: Alphabetical (ascii) order:
- `FunctionDeclaration` before `FunctionExpression`

**Review by acjh — CHANGES_REQUESTED** [review]

**Comment by acjh** [comment]
> Perhaps we should document this in the Wiki, together with the setup instructions.

Maybe *not* with the setup instructions (seems out-of-place in the [Developer Guide](https://github.com/MarkBind/markbind-cli/wiki/Developer-Guide)).

**Comment by nicholaschuayunzhi** [comment]
I'm working on the explanation in a separate gist i can link to in the dev guide. Sorry for missing out the previous ordering of keys.

**Comment by nicholaschuayunzhi** [comment]
I've removed ArrayExpression option in indent as the default option (1) account for aligning elements based on 2 spaces of indentation.

I've added the `"array-bracket-newline": ["error", { "multiline": true }],` to be consistent with airbnb's
`object-curly-newline` rule, as well as to prevent such code from passing which is not wanted based on nus oss js standard.
```
let array = [1,
  2,
  3,]
```

**Comment by nicholaschuayunzhi** [comment]
I've added a [gist](https://gist.github.com/nicholaschuayunzhi/bfe53dbb5f1a0f02d545d55956f9ab7e) for the explanation to the developer guide.

**Comment by acjh** [comment]
> to be consistent with airbnb's `object-curly-newline` rule

I don't see this 🤔

File: .eslintrc.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  "extends": "airbnb-base",
+  "rules": {
+    "array-bracket-newline": ["error", { "multiline": true }],
+    "func-names": "off",
+    "function-paren-newline": "off",
+    "indent": [
+      "error",
+      2,
+      {
+        "CallExpression": { "arguments": "first" } ,
```
Comment (acjh) [inline .eslintrc.js]: Remove space before comma.

**Review by acjh — CHANGES_REQUESTED** [review]

**Comment by acjh** [comment]
> I've added a gist for the explanation to the developer guide.

Nice. This can be a page in the Wiki. Let's [create a sidebar](https://help.github.com/articles/creating-a-sidebar/); the current pages need categorisation anyway.

**Comment by nicholaschuayunzhi** [comment]
> to be consistent with airbnb's object-curly-newline rule
>>I don't see this 🤔

Oh dear, i think i saw wrongly. But the rationale still stands to prevent the below from happening.
```
let array = [1,
  2,
  3,]
```

Weird though, my linter is firing off that error. Could it be a default rule? https://eslint.org/docs/rules/object-curly-newline

**Comment by acjh** [comment]
Both rules seem to be `{ "multiline": true }` by default.

**Comment by nicholaschuayunzhi** [comment]
Seems when i do not specify array-bracket-newline, it does not fire off any linting error.
image

Need to look into it further.

**Comment by nicholaschuayunzhi** [comment]
In the eslint docs for both rules, seems like [array-bracket-newline](https://eslint.org/docs/rules/array-bracket-newline) is not enabled by default.  
image



However for [object-curly-newline](https://eslint.org/docs/rules/object-curly-newline). It is enabled.
image

So i think we should still have the `array-bracket-newline` rule.

**Comment by acjh** [comment]
Nice catch.

**Comment by nicholaschuayunzhi** [comment]
I still find it weird though, according to [user guide](https://eslint.org/docs/rules/), no rules are enabled by default, but the docs seem to say otherwise.

**Comment by nicholaschuayunzhi** [comment]
Ok i think i got it, they disabled `array-bracket-newline` and enabled `object-curly-newline` in airbnb-base preset. And it is not documented on the readme.

https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb-base/rules/style.js

**Comment by nicholaschuayunzhi** [comment]
Okay I think it should be ready for review. Note that operator-linebreak is manually enabled as [airbnb-base package with enabled operator-linebreak rule is not published yet](https://github.com/airbnb/javascript/commit/6e1f11185d95432ba90a647efc027bd32983f17d)

I've also removed the line endings rule as airbnb-base already checks for unix endings

**Comment by Gisonrg** [comment]
LGTM, feel free to merge this @acjh

**Review by acjh — APPROVED** [review]

**Comment by acjh** [comment]
@nicholaschuayunzhi Thanks for your contribution!

**Comment by acjh** [comment]
@nicholaschuayunzhi Can you create a PR to apply `--fix`?

**Comment by nicholaschuayunzhi** [comment]
@acjh  Will do!

**Pull request merged**